File: packages/urdu-js/src/normalization/presentation.ts
import { PRESENTATION_FORMS_MAP } from './unicode-data.js'

export function normalizePresentationForms(text: string): string {
  if (!text) return text
  let result = ''
  for (const ch of text) {
    const cp = ch.codePointAt(0)!
    result += PRESENTATION_FORMS_MAP.get(cp) ?? ch
  }
  return result
}

File: packages/urdu-dart/lib/src/normalization/presentation.dart
import 'unicode_data.dart';

/// Converts Arabic Presentation Forms (U+FB50–FEFF) to their base characters.
String normalizePresentationForms(String text) {
  if (text.isEmpty) return text;
  final buf = StringBuffer();
  for (final r in text.runes) {
    final mapped = presentationFormsMap[r];
    if (mapped != null) {
      buf.write(mapped);
    } else {
      buf.writeCharCode(r);
    }
  }
  return buf.toString();
}
